src/udx.c

```diff
@@ -1187,38 +1176,46 @@ rack_detect_loss (udx_stream_t *stream) {
 
   int resending = 0;
   int mtu_probes_lost = 0;
+  udx_queue_node_t *p = NULL;
+  udx_queue_node_t *next = NULL; // save p->next so that we can remove it without breaking iteration
 
-  for (uint32_t seq = stream->remote_acked; seq != stream->seq; seq++) {
-    udx_packet_t *pkt = (udx_packet_t *) udx__cirbuf_get(&stream->outgoing, seq);
+  // can't use udx__queue_foreach because we may delete nodes in the middle
+  for (p = stream->inflight_queue.node.next, next = p->next;
+       p != &stream->inflight_queue.node;
+       p = next, next = p->next) {
 
-    if (pkt == NULL || pkt->lost) continue;
+    udx_packet_t *pkt = udx__queue_data(p, udx_packet_t, queue);
     assert(pkt->transmits > 0);
 
     // debug_printf("%lu > %lu=%d\n", stream->rack_time_sent, pkt->time_sent, stream->rack_time_sent > pkt->time_sent);
 
-    if (!rack_sent_after(stream->rack_time_sent, stream->rack_next_seq, pkt->time_sent, pkt->seq + 1)) {
-      continue;
+    if (pkt->time_sent > stream->rack_time_sent) {
```

Comment:
this is the optimization from RFC 8985:

> As an optimization, an implementation can choose to check only
   segments that have been sent before RACK.xmit_ts.  This can be more
   efficient than scanning the entire SACK scoreboard, especially when
   there are many segments in flight.  The implementation can use a
   separate doubly linked list ordered by Segment.xmit_ts, insert a
   segment at the tail of the list when it is (re)transmitted, and
   remove a segment from the list when it is delivered or marked as
   lost.  In Linux TCP, this optimization improved CPU usage by orders
   of magnitude during some fast recovery episodes on high-speed WAN
   networks.